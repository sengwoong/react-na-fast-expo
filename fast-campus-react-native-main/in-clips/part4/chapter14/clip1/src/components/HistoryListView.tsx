import React, {useCallback, useEffect, useState} from 'react';
import {FlatList, useWindowDimensions} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {getUserLikedHistory, TypeUserDispatch} from '../actions/user';
import {TypeDog} from '../data/TypeDog';
import {TypeUser} from '../data/TypeUser';
import {RootReducer} from '../store';
import {Button} from './Button';
import {RemoteImage} from './RemoteImage';
import database from '@react-native-firebase/database';

export const HistoryListView: React.FC<{
  onPressItem: (index: number) => void;
}> = props => {
  const {width} = useWindowDimensions();
  const dispatch = useDispatch<TypeUserDispatch>();

  //   const likedList = useSelector<RootReducer, TypeDog[]>(
  //     state => state.user.history,
  //   );
  const user = useSelector<RootReducer, TypeUser | null>(
    state => state.user.user,
  );
  const [likedList, setLikedList] = useState<TypeDog[]>([]);
  const loadLikedList = useCallback(async () => {
    // const user = getState().user.user;
    if (user === null) {
      return;
    }

    const ref = `history/${user.uid}`;
    const refCurrentHistory = await database().ref(ref);

    const currentHistory = await refCurrentHistory
      .once('value')
      .then(snapshot => snapshot.val());

    const dogList = Object.keys(currentHistory).map(key => {
      const item = currentHistory[key];

      return {
        photoUrl: item.url,
      } as TypeDog;
    });

    setLikedList(dogList);
  }, [user]);

  useEffect(() => {
    try {
      loadLikedList();
      //   dispatch(getUserLikedHistory());
    } catch (ex) {
      console.log(ex);
    }
  }, [dispatch, loadLikedList]);

  return (
    <FlatList<TypeDog>
      data={likedList}
      numColumns={2}
      renderItem={({item, index}) => (
        <Button
          testID={`Button${index}`}
          onPress={() => {
            //   setSelectedIdx(index);
            //   setIsVisible(true);
            props.onPressItem(index);
          }}>
          <RemoteImage
            width={width * 0.5}
            height={width * 0.5}
            url={item.photoUrl}
          />
        </Button>
      )}
    />
  );
};
