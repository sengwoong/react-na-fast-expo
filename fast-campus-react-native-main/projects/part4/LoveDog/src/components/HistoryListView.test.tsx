import {render, waitFor} from '@testing-library/react-native';
import {HistoryListView} from './HistoryListView';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from 'react-redux';

jest.mock('@react-native-firebase/database', () => {
  return () => ({
    ref: jest.fn().mockImplementation(() => ({
      once: jest.fn().mockReturnValue(
        Promise.resolve({
          val: jest.fn().mockReturnValue({
            TEST: {
              url: 'test',
            },
          }),
        }),
      ),
    })),
  });
});

describe('HistoryListView Render Test', () => {
  const initialState = {};
  const mockStore = configureStore([thunk]);

  //@ts-ignore
  let store;

  beforeAll(() => {
    store = mockStore({
      user: {
        user: {
          uid: 'TEST_ID',
        },
      },
    });
  });
  test('HistoryListView Render Snapshot Test', async () => {
    const onPressItem = jest.fn();

    const compoenent = render(
      //ts-ignore
      <Provider store={store}>
        <HistoryListView onPressItem={onPressItem} />
      </Provider>,
    );

    await waitFor(() => compoenent.findByTestId('Button0'), {
      timeout: 2000,
    });

    expect(compoenent.toJSON()).toMatchSnapshot();
  });
});
