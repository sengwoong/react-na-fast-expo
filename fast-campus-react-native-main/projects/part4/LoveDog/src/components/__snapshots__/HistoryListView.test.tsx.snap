// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryListView Render Test HistoryListView Render Snapshot Test 1`] = `
<RCTScrollView
  data={
    [
      {
        "photoUrl": "test",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": undefined,
              "paddingHorizontal": undefined,
              "paddingVertical": undefined,
            }
          }
          testID="Button0"
        >
          <Image
            source={
              {
                "uri": "test",
              }
            }
            style={
              [
                undefined,
                {
                  "height": 375,
                  "width": 375,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
